Source/AliveLibAE/Renderer/OpenGLRenderer.cpp

```diff
@@ -0,0 +1,1519 @@
+    { 1, 0, 0, 1.0f, 1.0f, 1.0f, 1, 0 },
+    { 1, 1, 0, 1.0f, 1.0f, 1.0f, 1, 1 },
+    { 0, 1, 0, 1.0f, 1.0f, 1.0f, 0, 1 } };
+
+    mTextureShader.Use();
+
+    Renderer_BindPalette(pPal);
+    Renderer_BindTexture(pTexture);
+
+    // Set our Projection Matrix, so stuff doesn't get rendered in the quantum realm.
+    mTextureShader.UniformMatrix4fv("m_MVP", GetMVP(sprt.mBase.vert.x, sprt.mBase.vert.y, sprt.field_14_w, sprt.field_16_h));
+
+    mTextureShader.Uniform1i("m_Sprite", 0); // Set m_Sprite to GL_TEXTURE0
+    mTextureShader.Uniform1i("m_Palette", 1); // Set m_Palette to GL_TEXTURE1
+    mTextureShader.Uniform1i("m_Textured", true);
+    mTextureShader.Uniform1i("m_PaletteEnabled", pPal != nullptr);
+
+    if (pPal != nullptr)
+        mTextureShader.Uniform1i("m_PaletteDepth", pPal->mPalDepth);
+
+    const GLuint indexData[6] = { 0, 1, 3, 3, 1, 2 };
+    DrawTriangles(verts, 4, indexData, 6);
+
+    mTextureShader.UnUse();
+}
+
+static GLuint TempGasEffectTexture = 0;
+
+void OpenGLRenderer::Draw(Prim_GasEffect& gasEffect)
+{
+    if (!gRenderEnable_GAS)
+        return;
+
+    if (TempGasEffectTexture == 0)
+        TempGasEffectTexture = Renderer_CreateTexture(GL_LINEAR);
+
+    if (gasEffect.pData == nullptr)
+        return;
+
+    int gasWidth = (gasEffect.w - gasEffect.x);
+    int gasHeight = (gasEffect.h - gasEffect.y);
+
+    glBindTexture(GL_TEXTURE_2D, TempGasEffectTexture);
+
+    glTexImage2D(GL_TEXTURE_2D, 0, GL_RGB565, gasWidth / 4, gasHeight / 2, 0, GL_RGB, GL_UNSIGNED_SHORT_5_6_5, gasEffect.pData);
+    
+    mTextureShader.Use();
+    mTextureShader.Uniform1i("m_Dithered", 1);
+    mTextureShader.Uniform1i("m_DitherWidth", gasWidth);
+    mTextureShader.Uniform1i("m_DitherHeight", gasHeight);
+    Renderer_SetBlendMode(TPageAbr::eBlend_1);
+    DrawTexture(TempGasEffectTexture, (float)gasEffect.x, (float)gasEffect.y, (float)gasWidth, (float)gasHeight);
+    mTextureShader.Use();
+    mTextureShader.Uniform1i("m_Dithered", 0);
+}
+
+void OpenGLRenderer::Draw(Prim_Tile& tile)
+{
+    if (!gRenderEnable_TILE)
+        return;
+
+    // todo: texturing ?
+    float r = tile.mBase.header.rgb_code.r / 255.0f;
+    float g = tile.mBase.header.rgb_code.g / 255.0f;
+    float b = tile.mBase.header.rgb_code.b / 255.0f;
+
+    VertexData verts[4] = {
+    { 0, 0, 0,  r, g, b,    0, 0 },
+    { 1, 0, 0,  r, g, b,    1, 0 },
+    { 1, 1, 0,  r, g, b,    1, 1 },
+    { 0, 1, 0,  r, g, b,    0, 1 } };
+
+    mTextureShader.Use();
+
+    mTextureShader.UniformMatrix4fv("m_MVP", GetMVP(tile.mBase.vert.x, tile.mBase.vert.y, tile.field_14_w, tile.field_16_h));
+    mTextureShader.Uniform1i("m_PaletteEnabled", false);
+    mTextureShader.Uniform1i("m_Textured", false);
+
+    glActiveTexture(GL_TEXTURE0);
+    glBindTexture(GL_TEXTURE_2D, 0);
+
+    const GLuint indexData[6] = { 0, 1, 3, 3, 1, 2 };
+    DrawTriangles(verts, 4, indexData, 6);
+
+    mTextureShader.UnUse();
+}
+
+void OpenGLRenderer::Draw(Line_F2& line)
```

Comment:
there is a lot of duplication in these Draw funcs, I can probably refactor that later though